File: api/schemas/update_user_profile/update_user_profile.py
import graphene

from graphql import GraphQLError

from app import logger
from db import db_session
from enums.languages import LanguageEnums
from functions.auth_wrappers import require_token
from functions.input_validators import cleanse_input
from models import Users
from scalars.email_address import EmailAddress


class UpdateUserProfileInput(graphene.InputObjectType):
    """
    This input object contains various non-required fields for use when updating
    user profile information
    """

    display_name = graphene.String(
        required=False, description="The new display name the user wishes to change to."
    )
    user_name = EmailAddress(
        required=False, description="The new user name the user wishes to change to."
    )
    preferred_lang = LanguageEnums(
        required=False,
        description="The new preferred language the user wishes to change to.",
    )
    password = graphene.String(
        required=False, description="The new password the user wishes to change to."
    )
    confirm_password = graphene.String(
        required=False, description="A password confirmation of their new password."
    )


class UpdateUserProfile(graphene.Mutation):
    """
    A mutation that allows the user to update their user profile to change various
    details of their current profile.
    """

    class Arguments:
        input = UpdateUserProfileInput(
            required=True,
            description="Various input fields, used in the updateUserProfile mutation.",
        )

    status = graphene.String(
        description="The status if the user profile update was successful."
    )

    @require_token
    def mutate(self, info, **kwargs):
        """
        This function will take in various arguments, and update the various parts
        of the user profile if included in the arguments
        :param info: Request Information
        :param kwargs: Various arguments passed in from the user
        :return: UpdateUserProfile with status field set
        """
        user_id = kwargs.get("user_id")
        display_name = cleanse_input(kwargs.get("input", {}).get("display_name", ""))
        user_name = cleanse_input(kwargs.get("input", {}).get("user_name", ""))
        preferred_lang = cleanse_input(
            kwargs.get("input", {}).get("preferred_lang", "")
        )
        password = cleanse_input(kwargs.get("input", {}).get("password", ""))
        confirm_password = cleanse_input(
            kwargs.get("input", {}).get("confirm_password", "")
        )

        # Get Current User
        user = db_session.query(Users).filter(Users.id == user_id).first()

        logging_str = ""

        if display_name != "":
            user.display_name = display_name
            logging_str += " display_name,"

        if user_name != "":
            user.user_name = user_name
            logging_str += " user_name,"

        if preferred_lang != "":
            user.preferred_lang = preferred_lang
            logging_str += " preferred_lang,"

        if password != "" and confirm_password != "" and password == confirm_password:
            user.password = password
            logging_str += " password,"
        if password != "" and confirm_password != "" and password != confirm_password:
            logger.warning(
                f"User: {user_id} attempted to update their password, however they do not match."
            )
            raise GraphQLError(
                "Error, passwords do not match. Unable to update profile."
            )

        try:
            # db_session.add(user)
            db_session.commit()
            logger.info(
                f"User: {user_id} successfully updated{logging_str} of their user profile."
            )
            return UpdateUserProfile(status="Successfully updated account.")
        except Exception as e:
            logger.error(
                f"Error occurred when user: {user_id} tried updating their profile."
            )
            raise GraphQLError(
                "Error when updating user information, please try again."
            )

File: api-js/src/mutations/user/update-user-profile.js
const { CIPHER_KEY } = process.env
const crypto = require('crypto')
const { GraphQLString } = require('graphql')
const { mutationWithClientMutationId } = require('graphql-relay')
const { GraphQLEmailAddress, GraphQLPhoneNumber } = require('graphql-scalars')
const { t } = require('@lingui/macro')
const { LanguageEnums } = require('../../enums')

const updateUserProfile = new mutationWithClientMutationId({
  name: 'UpdateUserProfile',
  description:
    'This mutation allows the user to update their user profile to change various details of their current profile.',
  inputFields: () => ({
    displayName: {
      type: GraphQLString,
      description: 'The updated display name the user wishes to change to.',
    },
    userName: {
      type: GraphQLEmailAddress,
      description: 'The updated user name the user wishes to change to.',
    },
    preferredLang: {
      type: LanguageEnums,
      description:
        'The updated preferred language the user wishes to change to.',
    },
    phoneNumber: {
      type: GraphQLPhoneNumber,
      description: 'The updated phone number the user wishes to change to.',
    },
  }),
  outputFields: () => ({
    status: {
      type: GraphQLString,
      description: 'The status if the user profile update was successful.',
      resolve: async (payload) => {
        return payload.status
      },
    },
  }),
  mutateAndGetPayload: async (
    args,
    {
      i18n,
      query,
      userKey,
      loaders: { userLoaderByKey },
      validators: { cleanseInput },
    },
  ) => {
    // Cleanse Input
    const displayName = cleanseInput(args.displayName)
    const userName = cleanseInput(args.userName).toLowerCase()
    const preferredLang = cleanseInput(args.preferredLang)
    const phoneNumber = cleanseInput(args.phoneNumber)

    // Make sure userKey is not undefined
    if (typeof userKey === 'undefined') {
      console.warn(
        `User attempted to update their profile, but the user id is undefined.`,
      )
      throw new Error(i18n._(t`Authentication error, please sign in again.`))
    }

    // Get user info from DB
    const user = await userLoaderByKey.load(userKey)

    if (typeof user === 'undefined') {
      console.warn(
        `User: ${userKey} attempted to update their profile, but no account is associated with that id.`,
      )
      throw new Error(i18n._(t`Unable to update profile. Please try again.`))
    }

    let updatedPhoneDetails, phoneValidated
    if (user.phoneValidated && typeof phoneNumber !== 'undefined') {
      const { iv, tag, phoneNumber: encryptedData } = user.phoneDetails
      const decipher = crypto.createDecipheriv(
        'aes-256-ccm',
        String(CIPHER_KEY),
        Buffer.from(iv, 'hex'),
        { authTagLength: 16 },
      )
      decipher.setAuthTag(Buffer.from(tag, 'hex'))
      let decrypted = decipher.update(encryptedData, 'hex', 'utf8')
      decrypted += decipher.final('utf8')

      if (decrypted !== phoneNumber) {
        updatedPhoneDetails = {
          iv: crypto.randomBytes(12).toString('hex'),
        }
        const cipher = crypto.createCipheriv(
          'aes-256-ccm',
          String(CIPHER_KEY),
          Buffer.from(updatedPhoneDetails.iv, 'hex'),
          { authTagLength: 16 },
        )
        let encrypted = cipher.update(phoneNumber, 'utf8', 'hex')
        encrypted += cipher.final('hex')

        updatedPhoneDetails.phoneNumber = encrypted
        updatedPhoneDetails.tag = cipher.getAuthTag().toString('hex')
      } else {
        phoneValidated = true
        updatedPhoneDetails = user.phoneDetails
      }
    }

    // Create object containing updated data
    const updatedUser = {
      displayName: displayName || user.displayName,
      userName: userName || user.userName,
      preferredLang: preferredLang || user.preferredLang,
      phoneDetails: updatedPhoneDetails,
      phoneValidated: phoneValidated || user.phoneValidated,
    }

    try {
      await query`
        UPSERT { _key: ${user._key} }
          INSERT ${updatedUser}
          UPDATE ${updatedUser} 
          IN users
      `
    } catch (err) {
      console.error(
        `Database error ocurred when user: ${user._key} attempted to update their profile: ${err}`,
      )
      throw new Error(i18n._(t`Unable to update profile. Please try again.`))
    }

    console.info(`User: ${user._key} successfully updated their profile.`)
    return {
      status: i18n._(t`Profile successfully updated.`),
    }
  },
})

module.exports = {
  updateUserProfile,
}
